### PR TITLE
serde for nodestore

### PR DESCRIPTION
**Pull request by sudiptoguha**
*Description of changes:* Mappers etc for the new trees (along with conversion from 2.0/2.1 versions for double precision). The next step would be deletion and deprecation of double precision.

### CODE REVIEW

File: Java/core/src/main/java/com/amazon/randomcutforest/CommonUtils.java [hunk 1]
```diff
@@ -200,7 +200,7 @@ public static double defaultScalarNormalizerFunction(double scalarValue, double
     };
 
     public static double[] toDoubleArray(float[] point) {
-        checkNotNull(point, "point must not be null");
+        checkNotNull(point, "array must not be null");
```
Comment (jotok) [inline Java/core/src/main/java/com/amazon/randomcutforest/CommonUtils.java:203]: The error message was written to match the formal parameter name, so I would either change this back or change the formal parameter to `array`.

File: Java/core/src/main/java/com/amazon/randomcutforest/sampler/CompactSampler.java [hunk 2]
```diff
@@ -219,6 +219,12 @@ public void reheap(boolean validate) {
         }
     }
 
+    public void addPoint(Integer pointIndex, float weight, long sequenceIndex) {
```
Comment (jotok) [inline Java/core/src/main/java/com/amazon/randomcutforest/sampler/CompactSampler.java:222]: Why did this method get added now? What changed?

File: Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java [hunk 3]
```diff
@@ -823,6 +823,10 @@ public boolean pointEquals(int index, float[] point) {
         int address = getLocation(index);
 
         if (!rotationEnabled) {
+
+              if (store.length < (address + dimensions)) {
+                  System.out.println("AAA " + store.length + " " + address + " " + (address + dimensions));
```
Comment (jotok) [inline Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java:828]: Remove

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/NewRandomCutTree.java [hunk 4]
```diff
@@ -58,38 +59,42 @@
...
 
     private Random testRandom;
+    protected boolean storeSequenceIndexesEnabled;
+    protected boolean centerOfMassEnabled;
     private long randomSeed;
     protected int root;
     protected IPointStoreView<float[]> pointStoreView;
-    protected int treeMass;
-    protected int capacity;
+    // protected int treeMass;
```
Comment (jotok) [inline Java/core/src/main/java/com/amazon/randomcutforest/tree/NewRandomCutTree.java:67]: Remove?

File: Java/core/src/main/java/com/amazon/randomcutforest/sampler/CompactSampler.java (hunk 2 above)
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/sampler/CompactSampler.java:222]: This is to enable double[] -> float[] conversion for the trees. Up until now we were bringing up the exact same samplers in ser-de. The guarantee we maintained is that for duplicates in the same tree, the corresponding sampler would have the same pointIndex value. In the conversion, we first round off the pointstore -- but then we may be creating duplicates. So we have to insert the points into a fresh tree and update the pointIndex. One option is to add the point to the samplerPlusTree and update both sampler and tree -- we could use the existing code. Other option is to add them to tree but not affect the sampler (other than updating the pointIndex value) -- which is happening here. Advantage of option 2 is that the sampler effectively remains in the same state; the scores seem a bit smoother in the conversion. Looking far down the road, this should also enable adding "sticky"  points to a sampler -- but that would require a resize() for grow() option to dynamically change (increase is the difficult part) the sampler size.

File: Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java (hunk 3 above)
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java:828]: Done

File: Java/core/src/main/java/com/amazon/randomcutforest/tree/NewRandomCutTree.java (hunk 4 above)
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/tree/NewRandomCutTree.java:67]: Done

File: Java/core/src/main/java/com/amazon/randomcutforest/CommonUtils.java (hunk 1 above)
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/CommonUtils.java:203]: Done

File: Java/core/src/main/java/com/amazon/randomcutforest/state/store/PointStoreFloatMapper.java [hunk 5]
```diff
@@ -50,7 +53,27 @@ public PointStore toModel(PointStoreState state, long seed) {
         int[] refCount = ArrayPacking.unpackInts(state.getRefCount(), indexCapacity, state.isCompressed());
         int[] locationList = new int[indexCapacity];
         int[] tempList = ArrayPacking.unpackInts(state.getLocationList(), state.isCompressed());
-        System.arraycopy(tempList, 0, locationList, 0, tempList.length);
+        if (!state.getVersion().equals(Version.V3_0)) {
```
Comment (jotok) [inline Java/core/src/main/java/com/amazon/randomcutforest/state/store/PointStoreFloatMapper.java:56]: If we start having to create more branching logic like this, I suggest we start creating separate methods for clarity:
```
public PointStore toModel(PointStoreState state, long seed) {
  if (state.getVersion.compareTo(Version.V3_0) < 0) {
    return toModelV2(state, seed);
  } else {
    return toModelV3(state, seed);
  }
}

private PointStore toModelV2(PointStoreState state, long seed) { ... }
private PointStore toModelV3(PointStoreState state, long seed) { ... }
```
}
```

**Review by jotok — APPROVED** [review]

File: Java/core/src/main/java/com/amazon/randomcutforest/state/store/PointStoreFloatMapper.java (hunk 5 above)
Comment (sudiptoguha) [reply Java/core/src/main/java/com/amazon/randomcutforest/state/store/PointStoreFloatMapper.java:56]: Agree wholeheartedly. Lets even start now - while we remember changes. Will push changes in the next few PRs (maybe keep the next one focused on just delete).

**Pull request merged**